collector/generatorreceiver/internal/generator/trace_generator.go

```diff
@@ -110,25 +110,34 @@ func (g *TraceGenerator) createSpanForServiceRouteCall(traces *pdata.Traces, ser
 			}
 		}
 	}
-	maxEndTime := startTimeMicros
+	maxChildEndTime := startTimeNanos
 	for s, r := range route.DownstreamCalls {
-		var childStartTimeMicros int64
+		var childStartTimeNanos = startTimeNanos
 		if route.LatencyPercentiles != nil {
-			childStartTimeMicros = startTimeMicros + int64(route.LatencyPercentiles.Sample())
+			childStartTimeNanos += route.LatencyPercentiles.Sample()
 		} else {
-			childStartTimeMicros = startTimeMicros + (g.random.Int63n(route.MaxLatencyMillis * 1000000))
+			childStartTimeNanos += g.random.Int63n(route.MaxLatencyMillis * 1000000)
 		}
 		childSvc := g.topology.GetServiceTier(s)
 
-		g.createSpanForServiceRouteCall(traces, childSvc, r, childStartTimeMicros, traceId, newSpanId)
-		maxEndTime = Max(maxEndTime, childStartTimeMicros)
+		childSpan := g.createSpanForServiceRouteCall(traces, childSvc, r, childStartTimeNanos, traceId, newSpanId)
+		maxChildEndTime = Max(maxChildEndTime, int64(childSpan.EndTimestamp()))
 	}
 
 	// todo: ownDuration should also be influenced by percentiles or not?
 	// note - changing this number seems to effect very little
-	ownDuration := g.random.Int63n(route.MaxLatencyMillis * 1000000)
-	span.SetStartTimestamp(pdata.NewTimestampFromTime(time.Unix(0, startTimeMicros)))
-	span.SetEndTimestamp(pdata.NewTimestampFromTime(time.Unix(0, maxEndTime+ownDuration)))
+	maxLatency := route.MaxLatencyMillis * 1000000
+	if route.LatencyPercentiles != nil {
+		maxLatency = route.LatencyPercentiles.Sample()
+	}
```

Comment:
I think given how this is used now it should probably be moved into the `if len(route.DownstreamCalls) == 0` block below and renamed "myLatency" or the like. It probably doesn't make sense to do the `g.random.Int63n(maxLatency)` if we're sampling from percentiles, either, since that will artificially lower the latency value. Hm, and now that I've written that I see a bit more simplification I'd like to do. I'll add a commit :)

Reply:
(did that and addressed a few other little comments I had)